### PR TITLE
Implement bitwise operations on integers

### LINKED ISSUE #3097: Bitwise operations on integers
### Description

It would be useful to add bitwise operations for integers to the `calc` module, i.e.,
- `not`, `and`, `or`, and `xor` as well as
- `lshift`, and `rshift`.

I am not sure if someone has already suggested this since it is very obvious but could not find any issue related to this. 

### Use Case

Bitwise operations have many applications and are in particular part of many widely-used algorithms. Although Typst is not a general-purpose programming language, I believe there will be sufficient use cases to justify including them. 

Another argument is that it is very hard and inefficient to implement bitwise operations with the available standard arithmetic functions (bit shift is not too difficult but the others are), so there would be no workaround. 

I **don't** suggest dedicated syntax (like `&` or `<<` for bitwise operations) because they will not be *that* common. Rather, I think that functions like `calc.and(a,b)` etc. would be appropriate.

### PR DESCRIPTION
**Pull request by PgBiel**
Closes #3097

Adds the following integer methods:
- `int.bit-not(x)`
- `int.bit-and(a, b)`
- `int.bit-or(a, b)`
- `int.bit-xor(a, b)`
- `int.bit-lshift(x, s)`
- `int.bit-rshift(x, s)`
  - `int.bit-rshift(x, s, logical: true)` is also implemented for logical right shift (doesn't extend the sign bit).

I deliberately made the right shift functions not error (the second parameter, `s`, is capped at 63, so it just "stops shifting" when it reached the limit, saturating at 0 for non-negative numbers and at -1 for negative numbers). This behavior is debatable (Rust panics when the shift is larger than `bits - 1` for example), but other languages closer to Typst seem to do the same (such as Python, and even JavaScript - although JavaScript's right shift seems to be a bit buggy cuz `Number` is a weird thing, but whatever).

The function names are bikesheddable (in particular the shift ones).

### CODE REVIEW

File: crates/typst/src/foundations/calc.rs [hunk 1]
```diff
@@ -459,6 +465,149 @@ pub fn ln(
...
+    #[default(false)]
+    logical: bool,
+) -> i64 {
+    // Don't panic on excessive shift.
+    let shift = shift.min(i64::BITS - 1);
+
+    if !logical || operand >= 0 {
+        // For non-negative operands, arithmetic and logical shifts produce the
+        // same results.
+        operand >> shift
```
Comment (frozolotl) [inline crates/typst/src/foundations/calc.rs:598]: I think the `operand >= 0` condition is unnecessary and confusing when reading the code. I'd suggest simply performing the unsigned shift when `logical` is true, no matter what.

File: crates/typst/src/foundations/calc.rs [hunk 2]
```diff
@@ -459,6 +465,149 @@ pub fn ln(
...
+    /// instead of arithmetic right shift (the default).
+    /// If this is `true`, negative operands will not preserve their sign bit,
+    /// and bits which appear to the left after the shift will be 0.
+    /// This parameter has no effect on non-negative operands.
+    #[named]
+    #[default(false)]
+    logical: bool,
+) -> i64 {
+    // Don't panic on excessive shift.
+    let shift = shift.min(i64::BITS - 1);
```
Comment (frozolotl) [inline crates/typst/src/foundations/calc.rs:593]: IMO we should perform a checked right shift, like is already done with the left shift. This increases consistency with that function, as well as all the other operations which also perform checked operations.

Comment (PgBiel) [reply crates/typst/src/foundations/calc.rs:593]: What I wanted initially was something like `saturating_shr`, but it doesn't seem to exist in Rust. So the next best thing was using a checked right shift, but to avoid a panic I'd have to provide a default value (`.unwrap_or(something)`). I realized that the default value would just be the result of capping the shift like that. So that's what I did.

Comment (PgBiel) [reply crates/typst/src/foundations/calc.rs:593]: Regarding throwing an error or not, personally I think it's just not very useful to error if the shift is too large here. The result would be the same regardless.

Comment (frozolotl) [reply crates/typst/src/foundations/calc.rs:593]: What I mean is that an error would be thrown when `checked_shr` fails.

File: crates/typst/src/foundations/calc.rs (hunk 1 above)
Comment (PgBiel) [reply crates/typst/src/foundations/calc.rs:598]: When I was writing this for the first time, I wanted to avoid unnecessary extra logic. But turns out that the extra logic is just casting, which is a no-op, so it should be fine to remove this check.

File: crates/typst/src/foundations/calc.rs (hunk 2 above)
Comment (PgBiel) [reply crates/typst/src/foundations/calc.rs:593]: Yeah.. I just think it should be fine for Typst to be "smart" here, considering that `x >> y` where `y > 63` can be written as `(x >> y mod 63) >>^(quot(y, 63)) 63`, where `a >>^n b` indicates repeating the `a >> b` operation `n` times. (Those extra repetitions are, generally, useless though.)

The only problem with my approach, currently, is that a logical shift of `-1` will stop at `1` for `shift >= 63`, when it should stop at `0` when `shift > 63` (if we want said smart behavior above).

File: crates/typst/src/foundations/calc.rs (hunk 1 above)
Comment (PgBiel) [reply crates/typst/src/foundations/calc.rs]: Done, thanks for the suggestion!

File: crates/typst/src/foundations/calc.rs (hunk 2 above)
Comment (frozolotl) [reply crates/typst/src/foundations/calc.rs:593]: There are three possible behaviors for an overflowing shift right operation:
- Wrap around, i.e. perform `operand >> (shift % 64)` (as done by JS)
- Set the result to zero or negative one, depending on signedness (as done by C)
- Error (as done by Rust)

To me, erroring seems like the most reasonable behavior, mainly because wrong assumptions will fail early.
I would be fine with the "zeroing" behavior too though, as long as it really behaves like it does in C. As it is right now, it seems incorrect as you already pointed out.

Comment (PgBiel) [reply crates/typst/src/foundations/calc.rs]: I've fixed the code to properly saturate at 0 now, like C.

I'm thinking here that this behavior allows the right shift to function regardless of amount of bits in the internal representation of integers, so this might make more sense for Typst overall as a more high-level language. Left shift is just a bit unfortunate in this regard; same as `bit-not`.

Although maybe we could allow users to pick less bits for `bit-not`, or choose signed / unsigned? That would probably be better suited for a future PR though.

**Comment by PgBiel** [comment]
Btw, I considered adding support for bitwise rotation operations (which are also supported natively by Rust), but I believe that could be out of scope for this PR (which aims to add just the most basic operations suggested in the linked issue), so I've omitted them for now. Perhaps they could be added in a future PR though, but we could discuss that in a separate issue.

**Comment by laurmaedje** [comment]
Could you summarize how signedness affects these operations? Are there cases in which `a op b` yields different results that `((a as u64) op (b as u64)) as i64` for any of these operations?

**Comment by Enter-tainer** [comment]
IIRC the only difference is right shift.

**Comment by PgBiel** [comment]
> Could you summarize how signedness affects these operations? Are there cases in which `a op b` yields different results that `((a as u64) op (b as u64)) as i64` for any of these operations?

According to the Rust reference, i64<->u64 is a no-op in terms of binary representation, so operations which only consider bits and not sign shouldn't differ there. Therefore, I believe only right shift will yield different results, since it will extend the sign bit by default (that's also why I added a parameter to toggle signed and unsigned right shift).

Here are some examples: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=fe5f97944c06c5f8c00a2f21359dbfdd

**Comment by laurmaedje** [comment]
Based on https://github.com/typst/typst/pull/3118#issuecomment-1889187552, we should probably migrate these to int methods.

**Comment by PgBiel** [comment]
Migration is done :+1:

**Comment by laurmaedje** [comment]
Thanks!
